tools/ci-cdk/canary-runner/src/build_bundle.rs

```diff
@@ -0,0 +1,462 @@
+    }
+    write!(output, "\n[features]\n").unwrap();
+    for release_tag in NOTABLE_SDK_RELEASE_TAGS.iter() {
+        writeln!(
+            output,
+            "\"{release_tag}\" = []",
+            release_tag = release_tag.as_str().replace('-', "_")
+        )
+        .unwrap();
+    }
+    writeln!(
+        output,
+        "default = [{enabled}]",
+        enabled = enabled_features(&crate_source)
+            .into_iter()
+            .map(|f| format!("\"{f}\""))
+            .collect::<Vec<String>>()
+            .join(", ")
+    )
+    .unwrap();
+    Ok(output)
+}
+
+fn sha1_file(path: &Path) -> Result<String> {
+    use sha1::{Digest, Sha1};
+    let contents = fs::read(path).context(here!())?;
+    let mut hasher = Sha1::new();
+    hasher.update(&contents);
+    Ok(hex::encode(hasher.finalize()))
+}
+
+pub async fn build_bundle(opt: BuildBundleOpt) -> Result<Option<PathBuf>> {
+    let canary_path = opt
+        .canary_path
+        .unwrap_or_else(|| std::env::current_dir().expect("current dir"));
+
+    // Determine the SDK crate source from CLI args
+    let crate_source = match (opt.sdk_path, opt.sdk_release_tag) {
+        (Some(sdk_path), None) => CrateSource::Path(sdk_path),
+        (None, Some(release_tag)) => CrateSource::VersionsManifest {
+            versions: VersionsManifest::from_github_tag(&release_tag).await?,
+            release_tag,
+        },
+        _ => unreachable!("clap should have validated against this"),
+    };
+
+    // Generate the canary Lambda's Cargo.toml
+    let manifest_path = canary_path.join("Cargo.toml");
+    let crate_manifest_content = generate_crate_manifest(crate_source)?;
+    fs::write(&manifest_path, crate_manifest_content).context("failed to write Cargo.toml")?;
+
+    if !opt.manifest_only {
+        // Compile the canary Lambda
+        let mut command = Command::new("cargo");
+        command
+            .arg("build")
+            .arg("--release")
+            .arg("--manifest-path")
+            .arg(&manifest_path);
+        if opt.musl {
+            command.arg("--target=x86_64-unknown-linux-musl");
+        }
+        handle_failure("cargo build", &command.output()?)?;
+
+        // Bundle the Lambda
+        let repository_root = find_git_repository_root("smithy-rs", canary_path)?;
+        let target_path = {
+            let mut path = repository_root.join("tools").join("target");
+            if opt.musl {
+                path = path.join("x86_64-unknown-linux-musl");
+            }
+            path.join("release")
+        };
+        let bin_path = target_path.join("bootstrap");
+        let bin_hash = sha1_file(&bin_path)?;
+        let bundle_path = target_path.join(format!("canary-lambda-{bin_hash}.zip"));
+
+        let zip_file = fs::File::create(&bundle_path).context(here!())?;
+        let mut zip = zip::ZipWriter::new(zip_file);
+        zip.start_file(
+            "bootstrap",
+            zip::write::FileOptions::default().unix_permissions(0o755),
```

Comment:
We'll never run these canaries on a non-linux system right?

Reply:
That's correct.